### ISSUE TITLE
Posibility of updating VOD workflow to validate MediaFile

### ISSUE THREAD

**Issue by skawirat** [awslabs/video-on-demand-on-aws#60]
Hi Daniel,

On the weekend our production system was impaired by AWS MediaConvert failing on multiple VOD jobs. Even though the particular files  were submitted multiple time they still continued to failed over and over again while at the time other files seemed to get processed without a issue 

MC simply gave error code 1999 - Unknown error. We have opened case ID 6814048611 to investigate the issue.

We have no insights  of the internal implementation of AWS MediaConvert but suspecting  for whatever reason MC didn't like a batch of particular files. 

**Since your team have more insights to MC implementation would it be possible to update AWS VOD workflow such that files which might possibly cause MC to throw up error will get rejected by VOD ?**

P.S: If you can provide a secure way for us to upload the files I can provide you access to the  flles which caused AWS MediaConvert to error;  or alternatively if you can get access to the above case number the files have been uploaded there

Thanks

Sam

**Comment by dscpinheiro** [comment]
Hi @skawirat,

I'll check with the support team to get more details. If there's something we can do in the workflow to prevent files from causing unknown errors, we'll add it in a future release.

In the meantime, one thing you can do to troubleshoot is check the output for the _mediainfo_ function (available on the _/aws/lambda/stack-name-mediainfo_ CloudWatch log group), and compare the results from a file that processed successfully versus one that failed.

**Comment by skawirat** [comment]
Hi @dscpinheiro,

Thank you for your response.
AWS Support did get back and informed us the root cause of the issue was that the audio track for input file was corrupt. [Refer to FFmpeg out below.

Turns out MediaConvert uses tools such as Atom inspector, FFmpeg ..etc to sanity check the file. The main issue was MediaConvert sending us a error code of 1999 : Unknown error; as opposed to a indication that the file was corrupt leaving us in a situation not knowing why the MediaConvert job failed.

If one of the following can be implemented it would be great

1. AWS MediaConvert returned a more meaning full error 
2. AWS VOD solution performs the same sanity checks as MediaConvert and returns a error code

Regards

Sam


` ffmpeg -i  problem_file.mp4 -vn -f null -
ffmpeg version N-94530-g8cd96e13ee-tessus  https://evermeet.cx/ffmpeg/  Copyright (c) 2000-2019 the FFmpeg developers
  built with Apple LLVM version 10.0.1 (clang-1001.0.46.4)
  configuration: --cc=/usr/bin/clang --prefix=/opt/ffmpeg --extra-version=tessus --enable-avisynth --enable-fontconfig --enable-gpl --enable-libaom --enable-libass --enable-libbluray --enable-libdav1d --enable-libfreetype --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libmysofa --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopus --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-version3 --pkg-config-flags=--static --disable-ffplay
  libavutil      56. 33.100 / 56. 33.100
  libavcodec     58. 55.100 / 58. 55.100
  libavformat    58. 30.100 / 58. 30.100
  libavdevice    58.  9.100 / 58.  9.100
  libavfilter     7. 58.100 /  7. 58.100
  libswscale      5.  6.100 /  5.  6.100
  libswresample   3.  6.100 /  3.  6.100
  libpostproc    55.  6.100 / 55.  6.100
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x7fd05f000400] Sample size 1634992128 is too large
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'problem_file.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: mp42mp41
    creation_time   : 2020-02-14T22:37:23.000000Z
  Duration: 00:03:03.13, start: 0.000000, bitrate: 12309 kb/s
    Stream #0:0(eng): Video: h264 (Main) (avc1 / 0x31637661), yuv420p(tv, bt709), 1920x1080 [SAR 1:1 DAR 16:9], 9949 kb/s, 25 fps, 25 tbr, 25k tbn, 50k tbc (default)
    Metadata:
      creation_time   : 2020-02-14T22:37:23.000000Z
      handler_name    : ?Mainconcept Video Media Handler
      encoder         : AVC Coding
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 317 kb/s (default)
    Metadata:
      creation_time   : 2020-02-14T22:37:23.000000Z
      handler_name    : #Mainconcept MP4 Sound Media Handler
Stream mapping:
  Stream #0:1 -> #0:0 (aac (native) -> pcm_s16le (native))
Press [q] to stop, [?] for help
Output #0, null, to 'pipe:':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: mp42mp41
    encoder         : Lavf58.30.100
    Stream #0:0(eng): Audio: pcm_s16le, 48000 Hz, stereo, s16, 1536 kb/s (default)
    Metadata:
      creation_time   : 2020-02-14T22:37:23.000000Z
      handler_name    : #Mainconcept MP4 Sound Media Handler
      encoder         : Lavc58.55.100 pcm_s16le
Error while decoding stream #0:1: Invalid argument
[aac @ 0x7fd05f007a00] Multiple frames in a packet.
[aac @ 0x7fd05f007a00] Prediction is not allowed in AAC-LC.
Error while decoding stream #0:1: Invalid data found when processing input
size=N/A time=00:02:52.43 bitrate=N/A speed= 925x    
video:0kB audio:32328kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown`

**Comment by dscpinheiro** [comment]
Hi @skawirat,

Sorry for the delay in the response, the team was discussing what the best approach is.

We'll create a feature request with MediaConvert so that a better error message (or at least a different error code) is returned when the audio is corrupted, and also look at any possible checks we can include in the workflow.

**Comment by skawirat** [comment]
Thanks @dscpinheiro,

>We'll create a feature request with MediaConvert so that a better error message 
>(or at least a different error code) is returned when the audio is corrupted

To be a bit more generic here
As a general rule of thumb MediaConvert should be sending as much error information as possible whenever it encounters a issue (not just on a corrupt audio track)

The issue with unknown error is that end users are left in the dark without knowing what went wrong; and only option they have is to raise a support ticket with MediaConvert support team.

On the long run supporting/returning more error codes other than unknown error will actually cut down on time MediaConvert team have to spend answering customer queries as to what went wrong.

Regards,

Sam

**Comment by skawirat** [comment]
Hi @dscpinheiro,

To give you a update on above issue of corrupted files causing "unknown error" on AWS MediaConvert. To address the issue I looked at validating the files in our workflow; prior to them getting submitted to AWS VOD Solution.

For the validation I consider MediaInfo, FFMpeg, FfProbe, mplayer and exiftool.

From the above MediaInfo, exiftool did NOT detect the particular corruption that occurred.
FFMpeg did the most comprehensive validation out of all; but unfortunately it took a considerable time on large file (almost as close to the encoding time of the file) so we had compromise between thoroughness to the validation vs time to execute.

Finally due to flexibility of command line options ..etc out of the remaining two [mplayer & exiftool] we decided to use mplayer.

So in a nutshell we have tried to detect for file corruption prior to submitting jobs to MC; but they are not thorough as we would like.

My above findings kind of validates that its better for MediaConvert itself to return meaningful error code

Thanks

Sam

**Comment by dscpinheiro** [comment]
Thanks for the update @skawirat.

For the time being, we will most likely continue to use mediainfo (since it addresses most of our use cases and it does not have a significant impact on the workflow duration), but hopefully the information you shared is going to be helpful to other customers who want to use a different tool.
